src/mlir/design.md

```diff
@@ -3,69 +3,225 @@
-This approach seems very clean and sensible, and it could be used for accessing heap variables as well.
-
-For now, every function argument and local declaration (`let`, `var`) are allocated in the stack and its value stored there.
-This simplifies code generation (all variable symbols point to an address), but it doesn't help with value symbols and are really slow.
+We use the LLVM dialect for alloca/load/store/GEP operations when dealing with stack/heap objects.
 
-The LLVM pipeline, even at O1, recognises most of the argument, temporary and local allocas as bogus and optimises them away, so this isn't an immediately critical issue from a performance point of view, but we will want to clean up some of those repetitive patterns soon enough.
+Local declarations (`let`, `var`) without region information are allocated in the stack, via LLVM's `alloca` instruction and their types.
+Region objects are allocated using runtime calls to `snmalloc` with the right region types.
 
-## Runtime Calls
+This may be too low level, so it's a strong candidate to move up to a dialect operation.
+But we will only do so if there is a clear path for optimisation.
 
-The Verona runtime is where the magic happens.
-Allocating regions, choosing the memory management strategy, scheduling behaviours, dispatching foreign language code, etc.
-The AST, however, has only the semantics of the language (with hints to the runtime behaviour), so we need to translate it at the MLIR lowering level.
+## Arithmetic and Builtins
 
-For example, heap allocation, which LLVM IR generally uses calls to `@llvm.malloc`, Verona uses `snmalloc`, which has a finer control over what we can do with the regions.
-In the end, however, it will be the same: calls to the runtime library in IR.
+Arithmetic and builtins are implemented in Verona via direct calls to intrinsics and MLIR/LLMV operations by name (string).
+Each numeric class will have their operations defined in the standard library and the compiler will detect the call as _internal_ and create the appropriate intrinsic / operation.
+Builtins (number conversion, checks, cpu-specific instructions) will also be implemented in the standard Verona library.
 
-The generator knows what runtime calls are appropriate for each AST construct, and can insert those calls in the right places and make sure to declare the methods so that linking with the runtime library works.
+Adding two numbers (`a + b`) is the same as calling the method `+` on `a` over `b` (ex. `a.+(b)` or `+(a, b)`).
+These calls will be matched against the already implemented methods in those classes and behave like a normal call (dynamic or static).
 
-## Arithmetic
+These methods are small enough that the LLVM inliner pass will convert all calls into single operations.
+If any of them proves more complex than the inliner can handle, we can add the _always_inline_ attribute to them.
```

Comment:
It's probably worth doing this anyway.  We probably want to inline them even in debug builds (where we may only run the always-inline inliner) and we almost certainly want to run the always-inline inliner at the start of our LLVM pass pipeline to make sure that these things are inlined before we do anything else.  Most of the pre-inliner passes won't do very much good if they're run before they can see arithmetic.